src/Ocelot/Requester/HttpClientBuilder.cs

```diff
@@ -0,0 +1,136 @@
+            }
+
+            var timeout = downstreamRoute.QosOptions.TimeoutValue == 0
+                ? _defaultTimeout
+                : TimeSpan.FromMilliseconds(downstreamRoute.QosOptions.TimeoutValue);
+
+            _httpClient = new HttpClient(CreateHttpMessageHandler(handler, downstreamRoute))
+            {
+                Timeout = timeout,
+            };
+
+            _client = new HttpClientWrapper(_httpClient, downstreamRoute.ConnectionClose); // TODO
+
+            return _client;
+        }
+
+        private static HttpClientHandler CreateHandler(DownstreamRoute downstreamRoute)
+        {
+            // Dont' create the CookieContainer if UseCookies is not set or the HttpClient will complain
+            // under .Net Full Framework
+            var useCookies = downstreamRoute.HttpHandlerOptions.UseCookieContainer;
+
+            return useCookies ? UseCookiesHandler(downstreamRoute) : UseNonCookiesHandler(downstreamRoute);
+        }
+
+        private static HttpClientHandler UseNonCookiesHandler(DownstreamRoute downstreamRoute)
+        {
+            return new HttpClientHandler
+            {
+                AllowAutoRedirect = downstreamRoute.HttpHandlerOptions.AllowAutoRedirect,
+                UseCookies = downstreamRoute.HttpHandlerOptions.UseCookieContainer,
+                UseProxy = downstreamRoute.HttpHandlerOptions.UseProxy,
+                MaxConnectionsPerServer = downstreamRoute.HttpHandlerOptions.MaxConnectionsPerServer,
+                
+            };
+        }
+
+        private static HttpClientHandler UseCookiesHandler(DownstreamRoute downstreamRoute)
+        {
+            return new HttpClientHandler
+            {
+                AllowAutoRedirect = downstreamRoute.HttpHandlerOptions.AllowAutoRedirect,
+                UseCookies = downstreamRoute.HttpHandlerOptions.UseCookieContainer,
+                UseProxy = downstreamRoute.HttpHandlerOptions.UseProxy,
+                MaxConnectionsPerServer = downstreamRoute.HttpHandlerOptions.MaxConnectionsPerServer,
+                CookieContainer = new CookieContainer(),
+            };
+        }
+
+        public void Save()
+        {
+            _cacheHandlers.Set(_cacheKey, _client, TimeSpan.FromHours(24));
+        }
+
+        private HttpMessageHandler CreateHttpMessageHandler(HttpMessageHandler httpMessageHandler, DownstreamRoute request)
+        {
+            //todo handle error
+            var handlers = _factory.Get(request).Data;
+
+            handlers
+                .Select(handler => handler)
+                .Reverse()
+                .ToList()
+                .ForEach(handler =>
+                {
+                    var delegatingHandler = handler();
+                    delegatingHandler.InnerHandler = httpMessageHandler;
+                    httpMessageHandler = delegatingHandler;
+                });
+            return httpMessageHandler;
+        }
```

Comment:
```suggestion
           return handlers
        .Reverse()  // Reverse the sequence
        .Aggregate(httpMessageHandler, (currentHandler, handlerFunc) =>
        {
            var delegatingHandler = handlerFunc();
            delegatingHandler.InnerHandler = currentHandler;
            return delegatingHandler;
        });
            return httpMessageHandler;
        }
```

Reply:
Yeap! `ToList` was a bit overhead to get access to `ForEach` helper.
`Aggregate` helper is good, for sure because it is applicable for this case.
There is another coding life hack: calling `All` to process all elements of collection. 😃 

**But!...** Ray, this file is redundant one! I left this file during rebasing just to keep the author's changes: see `// TODO` 😉 
Both files are just container of new changes
- src/Ocelot/Requester/HttpClientBuilder.cs
- src/Ocelot/Requester/HttpClientWrapper.cs

[You may check](https://github.com/search?q=repo%3AThreeMammals%2FOcelot%20HttpClientBuilder&type=code), there are no these files in develop! 🙃 
This PR is not ready for reviews: it must be developed more ❗